File: python/variable-length-quantity/variable_length_quantity.py
from functools import reduce

# Encode a list of numbers into a list of bytes
def encode(numbers):
    # Iterate over each number in the input list and encode it using encode_single_number
    return [x for number in numbers for x in encode_single_number(number)]

# Encode a single number into a list of bytes
def encode_single_number(number):
    # Initialize a list to store the encoded bytes
    bytes_ = [number % 0x80]  # Get the least significant 7 bits of the number
    number //= 0x80  # Shift the number right by 7 bits

    # Continue encoding until the number becomes zero
    while number > 0x0:
        bytes_.append(number % 0x80 + 0x80)  # Set the most significant bit to 1 and append
        number //= 0x80  # Shift the number right by 7 bits

    # Reverse the list of bytes to maintain the correct order
    return reversed(bytes_)

# Decode a list of bytes into a list of numbers
def decode(bytes_):
    i = 0
    numbers = []

    # Iterate over the bytes to decode the numbers
    for j in range(len(bytes_)):
        # If the byte has the most significant bit clear or it's the last byte
        if bytes_[j] < 0x80 or j == len(bytes_) - 1:
            # Decode the number using decode_single_number and add it to the list
            numbers.append(decode_single_number(bytes_[i:j + 1]))
            i = j + 1  # Update the starting index for the next number

    return numbers

# Decode a list of bytes into a single number
def decode_single_number(bytes_):
    # Perform error checks for incomplete sequences and invalid lengths
    if bytes_ == [] or bytes_[-1] > 0x7F or len(bytes_) > 5 \
            or (len(bytes_) == 5 and bytes_[0] > 0x8F):
        raise ValueError("incomplete sequence")

    # Use functools.reduce to decode the bytes into a single number
    return reduce(lambda acc, x: (acc << 7) + x % 0x80, bytes_, 0)

File: javascript/variable-length-quantity/variable-length-quantity.js
export const encode = (value) => {
  let result = [];
  
  // Loop until value is processed completely
  while (value > 0) {
    // Get the lowest 7 bits of the value
    let byte = value & 0x7F;

    // If there are more bits to encode, set the continuation bit (MSB)
    if (value > 0x7F) byte |= 0x80;

    // Add the byte to the result array
    result.push(byte);

    // Shift value to the right by 7 bits
    value >>= 7;
  }

  // If no bytes were added, the value is 0, so add a single zero byte
  if (result.length === 0) result.push(0);

  // Return the array of bytes in reverse order
  return result.reverse();
};

export const decode = (bytes) => {
  let result = 0;
  let shift = 0;

  // Loop through each byte in the array
  for (let i = 0; i < bytes.length; i++) {
    // Extract the lowest 7 bits from the byte
    let byte = bytes[i] & 0x7F;

    // Accumulate the byte's value into the result
    result |= byte << shift;

    // Update the shift by 7 bits for the next iteration
    shift += 7;

    // Check if the continuation bit (MSB) is not set
    if ((bytes[i] & 0x80) === 0) {
      return result;
    }
  }

  throw new Error('Invalid variable length quantity');
};
